lib/actions/deploy.coffee

```diff
@@ -0,0 +1,159 @@
+		process.stdout.write("Uploaded #{percent.toFixed(1)}%")
+		console.log() if percent == 100
+	, timeout
+
+getBundleInfo = (options) ->
+	helpers = require('../utils/helpers')
+
+	helpers.getAppInfo(options.appName)
+	.then (app) ->
+		[app.arch, app.device_type]
+
+performUpload = (image, token, username, url, size, appName) ->
+	request = require('request')
+	url = url || process.env.RESINRC_RESIN_URL
+	post = request.post
+		url: getBuilderPushEndpoint(url, username, appName)
+		auth:
+			bearer: token
+		body: image
+
+	uploadToPromise(post, size)
+
+uploadToPromise = (request, size) ->
+	new Promise (resolve, reject) ->
+
+		handleMessage = (data) ->
+			data = data.toString()
+			if process.env.DEBUG
+				console.log("Received data: #{data}")
+
+			obj = JSON.parse(data)
+			if obj.type?
+				switch obj.type
+					when 'error' then reject(new Error("Remote error: #{obj.error}"))
+					when 'success' then resolve(obj.image)
+					when 'status' then console.log("Remote: #{obj.message}")
+					else reject(new Error("Received unexpected reply from remote: #{data}"))
+			else
+				reject(new Error("Received unexpected reply from remote: #{data}"))
+
+		request
+		.on('error', reject)
+		.on('data', handleMessage)
+
+		# Set up upload reporting
+		pushProgress(size, request)
+
+
+module.exports =
+	signature: 'deploy <appName> [image]'
+	description: 'Deploy a container to a resin.io application'
+	help: '''
+		Use this command to deploy and optionally build an image to an application.
+
+		Usage: deploy <appName> ([image] | --build [--source build-dir])
+
+		Note: If building with this command, all options supported by `resin build`
+		are also support with this command.
+
+		Examples:
+		$ resin deploy myApp --build --source myBuildDir/
+		$ resin deploy myApp myApp/myImage
+	'''
+	permission: 'user'
+	options: [
+		{
+			signature: 'build'
+			boolean: true
+			description: 'Build image then deploy'
+			alias: 'b'
+		},
+		{
+			signature: 'source'
+			parameter: 'source'
+			description: 'The source directory to use when building the image'
+			alias: 's'
+		}
+	]
+	action: (params, options, done) ->
+		_ = require('lodash')
+		tmp = require('tmp')
+		tmpNameAsync = Promise.promisify(tmp.tmpName)
+		resin = require('resin-sdk-preconfigured')
+
+		dockerUtils = require('../utils/docker')
+
+		# Ensure the tmp files gets deleted
+		tmp.setGracefulCleanup()
+
+		docker = dockerUtils.getDocker(options)
+		# Check input parameters
+		parseInput(params, options)
+		.then ([appName, build, context, imageName]) ->
+			tmpNameAsync()
+			.then (tmpPath) ->
+
+				# Setup the build args for how the build routine expects them
+				options = _.assign({}, options, { appName })
+				params = _.assign({}, params, { context })
+
+				if build
+					# Instead of an image, the user has specified to build an image
+					imagePromise = dockerUtils.runBuild(params, options, getBundleInfo)
+				else
+					imagePromise = Promise.resolve(imageName)
```

Comment:
This can be done as
```coffee
Promise.try ->
	if build
		# Instead of an image, the user has specified to build an image
 		dockerUtils.runBuild(params, options, getBundleInfo)
	else
		imageName
.then (imageName) ->